lib/mapnik_backend.js

```diff
@@ -226,52 +207,64 @@ MapnikSource.registerProtocols = function(tilelive) {
 // can be expected to be in `mapfile`. If this isn't successful, `callback` gets
 // an error as its first argument.
 MapnikSource.prototype._loadXML = function(callback) {
-    var source = this;
-    this._base = path.resolve(path.dirname(this._uri.pathname));
+    if (this._uri.pathname) {
+        this._base = path.resolve(path.dirname(this._uri.pathname));
+    }
 
     // This is a string-based map file. Pass it on literally.
     if (this._uri.xml) return callback(null, this._uri.xml);
 
     // Load XML from file.
-    fs.readFile(path.resolve(source._uri.pathname), 'utf8', function(err, xml) {
-        if (err) return callback(err);
-        callback(null, xml);
+    fs.readFile(path.resolve(this._uri.pathname), 'utf8', (err, xml) => {
+        callback(err, xml);
     });
 };
 
 // Create a new mapnik map object at `this.mapnik`. Requires that the mapfile
 // be localized with `this.localize()`. This can be called in repetition because
 // it won't recreate `this.mapnik`.
-MapnikSource.prototype._createPool = function(xml, callback) {
-    var source = this;
-    if (!this._pool) this._pool = Pool({
-        create: function(callback) {
-            var map = new mapnik.Map(source._uri.query.tileSize, source._uri.query.tileSize);
-            map.bufferSize = source._uri.query.bufferSize;
-            var opts = {strict: source._uri.strict || false, base: source._base + '/'};
-            try {
-                map.fromString(xml, opts, function(err, map) {
-                    if (err) return callback(err);
-                    return callback(null, map);
-                });
-            } catch (err) {
-                return callback(err);
-            }
+MapnikSource.prototype._createPool = function(xml) {
+    if (this._pool) return;
+
+    const opts = {
+        max: this._uri.query.poolSize || process.env.UV_THREADPOOL_SIZE || require('os').cpus().length
+    };
+    const factory = {
+        create: () => {
+            return new Promise((resolve, reject) => {
```

Comment:
I understand your motivations to use promises here, but if you take a deeper look at what you are doing here:
```
. _populateInfo() [callback]-> .create() [promise] -> map.fromString() [callback]
```
The function `_populateInfo`, which uses callbacks, calls to `factory.create` which creates a promise and calls  `map.fromString` which is the original function that returns its value via callbacks. So we are using two ways to handle asynchronous code (callbacks and promises) but we are not taking advantage of promises at all, we are adding complexity for free. Obviously, the future (present to be honest) to handle asynchronous code are `promises`. But we need to transition to them in the right way. For that, we need to wrap or, better, reimplement the original function that produces asynchronous code to return promises. For instance:
```
. _populateInfo() [callback]-> .create() [callback] -> map.fromString() [promise]
```
then
```
. _populateInfo() [callback]-> .create() [promise] -> map.fromString() [promise]
```
and finally:
```
. _populateInfo() [promise]-> .create() [promise] -> map.fromString() [promise]
```
In conclusion to begin to use promises we must start from the bottom, that means, we need to reimplement `node-mapnik` to return promises (we can also wrap node mapnik with some [library](https://github.com/CartoDB/cartonik/blob/master/src/map-renderer.js#L10-L19) to move from callbacks to promises). Do you accept the challenge?? 😈 

For this PR, I encourage you to not use promises, please use callbacks instead. We are doing the same by removing `step` in `Windshaft-cartodb`, we are replacing `step` by callbacks. There is an exception to this rule, if you need to do perform the same asynchronous operation multiple times, as you did with this [PR](https://github.com/CartoDB/Windshaft/pull/632/files) to fetch column names for all layers.

Reply:
> In conclusion to begin to use promises we must start from the bottom, that means, we need to reimplement node-mapnik to return promises (we can also wrap node mapnik with some library to move from callbacks to promises). Do you accept the challenge?? smiling_imp

With the requirement of maintaining compatibility with both callback and promises, that's pretty tough to do in a reasonable amount of time in node-mapnik. Wanna create a project? I add promises and you join the modules between node-mapnik and Windshaft :smile:

> For this PR, I encourage you to not use promises, please use callbacks instead.

`generic-pool` removed the callback interface. So I can either revert the update, wrap the `acquire` and `release` with callbacks or leave the promises. I think either 1 or 3, but 2 is just going backwards. What do you think?

Reply:
> generic-pool removed the callback interface.

Whoa, what a pity. if `generic-pool` has other features that we can be interested in, go for 3. I not, go for 1.

Reply:
I've reverted to 2.5.4. Also I reworked _createMetatileCache and _renderMetatile to avoid the need to use `nextTick()`. Can you have another look, please?